src/FixedDataTableCellGroup.js

```diff
@@ -76,25 +87,70 @@ var FixedDataTableCellGroupImpl = React.createClass({
     }, false);
 
     var currentPosition = 0;
-    for (var i = 0, j = columns.length; i < j; i++) {
+    var count = 0;
+
+    var newColumnsToRender = new Array(props.maxVisibleColumns);
+    var positions = new Array(props.maxVisibleColumns);
+
+    for (var i = 0; i < columns.length; i++) {
       var columnProps = columns[i].props;
       var recycable = columnProps.allowCellsRecycling && !isColumnReordering;
       if (!recycable || (
-            currentPosition - props.left <= props.width &&
-            currentPosition - props.left + columnProps.width >= 0)) {
-        var key = columnProps.columnKey || 'cell_' + i;
-        cells[i] = this._renderCell(
-          props.rowIndex,
-          props.rowHeight,
-          columnProps,
-          currentPosition,
-          key,
-          contentWidth,
-          isColumnReordering
-        );
+          currentPosition - props.left <= props.width &&
+          currentPosition - props.left + columnProps.width >= 0)) {
+        positions[i] = currentPosition;
+        newColumnsToRender[count++] = i;
       }
+
       currentPosition += columnProps.width;
     }
+
+    //TODO move this recycle logic into main state
+    const newColumnsSet = new Set(newColumnsToRender);
+    const oldColumnsSet = new Set(this._columnsToRender);
+    const indexes = [];
+
+    for (var i = props.maxVisibleColumns; i >= 0; i--) {
+      var column = this._columnsToRender[i];
+      if (!column || !newColumnsSet.has(column)) {
+        indexes.push(i);
+      }
+    }
+
+    newColumnsToRender.forEach((column) => {
+      if (!oldColumnsSet.has(column)) {
+        this._columnsToRender[indexes.pop()] = column;
+      }
+    });
+
+    var cellCount = 0;
+    this._columnsToRender.forEach((i) => {
+      var columnProps = columns[i].props;
+      var currentPosition = positions[i];
+      var key = columnProps.columnKey || 'cell_' + i;
+      this._staticCellArray[cellCount++] = this._renderCell(
+        props.rowIndex,
+        props.rowHeight,
+        columnProps,
+        currentPosition,
+        key,
+        contentWidth,
+        isColumnReordering,
+      );
+    });
+
+    while (cellCount < props.maxVisibleColumns) {
+      if (!this._staticCellArray[cellCount]) {
+        break;
+      }
+
+      this._staticCellArray[cellCount] = React.cloneElement(this._staticCellArray[cellCount], {
+        visible: false,          
```

Comment:
Why do we need to clone?  Is the advantage of having this that it avoids a removal of DOM node & remount later?